core/kernel/dt.c

```diff
@@ -473,6 +496,223 @@ void *get_secure_dt(void)
 }
 
 #if defined(CFG_EMBED_DTB)
+#ifdef CFG_DT_CACHED_NODE_INFO
+/* Reference to the FDT for which node information are cached */
+static const void *cached_node_info_fdt;
+
+/*
+ * struct cached_node - Cached information of a DT node
+ *
+ * @node_offset: Offset of the node in @cached_node_info_fdt
+ * @parent_offset: Offset of @node_offset parent node
+ * @address_cells: #address-cells property value of the parent node or 0
+ * @size_cells: #size-cells property value of the parent node or 0
+ * @phandle: Phandle associated to the node or 0 if none
+ */
+struct cached_node {
+	int node_offset;
+	int parent_offset;
+	int8_t address_cells;
+	int8_t size_cells;
+	uint32_t phandle;
+};
+
+/*
+ * struct dt_node_cache - Reference to cached information of DT nodes
+ * @array: Array of the cached node
+ * @count: Number of initialized cells in @array
+ * @alloced_array
+ */
+struct dt_node_cache {
+	struct cached_node *array;
+	size_t count;
+	size_t alloced_count;
+};
+
+static struct dt_node_cache dt_node_cache;
+
+static struct cached_node *find_cached_parent_node(const void *fdt,
+						   int node_offset)
+{
+	struct cached_node *cell = NULL;
+	size_t n = 0;
+
+	if (!cached_node_info_fdt || fdt != cached_node_info_fdt)
+		return NULL;
+
+	for (n = 0; n < dt_node_cache.count; n++)
+		if (dt_node_cache.array[n].node_offset == node_offset)
+			cell = dt_node_cache.array + n;
+
+	return cell;
+}
+
+int fdt_find_cached_parent_node(const void *fdt, int node_offset,
+				int *parent_offset)
+{
+	struct cached_node *cell = NULL;
+
+	cell = find_cached_parent_node(fdt, node_offset);
+	if (!cell)
+		return -FDT_ERR_NOTFOUND;
+
+	*parent_offset = cell->parent_offset;
+
+	return 0;
+}
+
+int fdt_find_cached_parent_reg_cells(const void *fdt, int node_offset,
+				     int *address_cells, int *size_cells)
+{
+	struct cached_node *cell = NULL;
+	int rc = 0;
+
+	cell = find_cached_parent_node(fdt, node_offset);
+	if (!cell)
+		return -FDT_ERR_NOTFOUND;
+
+	if (address_cells) {
+		if (cell->address_cells >= 0)
+			*address_cells = cell->address_cells;
+		else
+			rc = -FDT_ERR_NOTFOUND;
+	}
+
+	if (size_cells) {
+		if (cell->size_cells >= 0)
+			*size_cells = cell->size_cells;
+		else
+			rc = -FDT_ERR_NOTFOUND;
+	}
+
+	return rc;
+}
+
+int fdt_find_cached_node_phandle(const void *fdt, uint32_t phandle,
+				 int *node_offset)
+{
+	struct cached_node *cell = NULL;
+	size_t n = 0;
+
+	if (!cached_node_info_fdt || fdt != cached_node_info_fdt)
+		return -FDT_ERR_NOTFOUND;
+
+	for (n = 0; n < dt_node_cache.count; n++)
+		if (dt_node_cache.array[n].phandle == phandle)
+			cell = dt_node_cache.array + n;
+
+	if (!cell)
+		return -FDT_ERR_NOTFOUND;
+
+	*node_offset = cell->node_offset;
+
+	return 0;
+}
+
+static TEE_Result grow_cached_node_array(void)
```

Comment:
The name of this function would be accurate without the `if` check below. Now the function rather guarantees room for at least one more element.

Reply:
`realloc_cached_node_array()` would maybe be better.